src/ILLink.RoslynAnalyzer/XmlAnalyzer.cs

```diff
@@ -0,0 +1,114 @@
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create (s_moreThanOneValueForParameterOfMethod, s_errorProcessingXmlLocation);
+
+		private static readonly Regex _linkAttributesRegex = new (@"ILLink\.LinkAttributes.*\.xml");
+
+		public override void Initialize (AnalysisContext context)
+		{
+			context.EnableConcurrentExecution ();
+			context.ConfigureGeneratedCodeAnalysis (GeneratedCodeAnalysisFlags.ReportDiagnostics);
+
+			context.RegisterCompilationStartAction (context => {
+				// Report Diagnostics on malformed XML with additionalFileContext and register actions to resolve names with (CompilationStartAnalysisContext) context.
+				context.RegisterAdditionalFileAction (additionalFileContext => {
+					// Check if it's a LinkAttributes xml
+					if (_linkAttributesRegex.IsMatch (Path.GetFileName (additionalFileContext.AdditionalFile.Path))) {
+						if (additionalFileContext.AdditionalFile.GetText () is not SourceText text)
+							return;
+
+						if (!ValidateLinkAttributesXml (additionalFileContext, text))
+							return;
+
+						if (!context.TryGetValue (text, ProcessXmlProvider, out var xmlData) || xmlData is null)
+							return;
+						foreach (var root in xmlData) {
+							if (root is LinkAttributes.TypeNode typeNode) {
+								foreach (var duplicatedMethods in typeNode.Methods.GroupBy (m => m.Name).Where (m => m.Count () > 0)) {
+									additionalFileContext.ReportDiagnostic (Diagnostic.Create (s_moreThanOneValueForParameterOfMethod, null, duplicatedMethods.FirstOrDefault ().Name, typeNode.FullName));
+								}
+							}
+						}
+					}
+				});
+			});
+		}
+
+		private static XmlSchema GenerateLinkAttributesSchema ()
+		{
+			var assembly = Assembly.GetExecutingAssembly ();
+			using var schemaStream = assembly.GetManifestResourceStream ("ILLink.RoslynAnalyzer.ILLink.LinkAttributes.xsd");
+			using var reader = XmlReader.Create (schemaStream);
+			var schema = XmlSchema.Read (
+				reader,
+				null);
+			return schema;
+		}
+		private static readonly XmlSchema LinkAttributesSchema = GenerateLinkAttributesSchema ();
+
+		private static bool ValidateLinkAttributesXml (AdditionalFileAnalysisContext context, SourceText text)
+		{
+			var xmlStream = GenerateStream (text);
+			XDocument? document;
+			try {
+				document = XDocument.Load (xmlStream, LoadOptions.SetLineInfo | LoadOptions.PreserveWhitespace);
+			} catch (XmlException ex) {
+				context.ReportDiagnostic (Diagnostic.Create (s_errorProcessingXmlLocation, null, ex.Message));
+				return false;
+			}
+			XmlSchemaSet schemaSet = new XmlSchemaSet ();
+			schemaSet.Add (LinkAttributesSchema);
+			bool valid = true;
+			document.Validate (schemaSet, (sender, error) => {
+				context.ReportDiagnostic (Diagnostic.Create (s_errorProcessingXmlLocation, null, "ILLinkAttributes.xml': '" + "At line " + error.Exception.LineNumber + ": " + error.Message));
```

Comment:
I'm realizing now that the XML schema enforcement overlaps with a number of the trimmer warnings, so this `ReportDiagnostic` should probably be removed.
Related to that though, would it be better / possible to remove the diagnostics that can be caught by a schema validation and only include things like failing to resolve type names?

Reply:
I'm personally all for using the schema to validate the XML. It will depend on what the schema validation warnings look like - we should try not to give worse warnings than the linker (I'm guessing the schema validator will give better information if anything). We might want to map some of the validation errors to the existing error codes.

I'd suggest doing whatever is easiest for now - since the more important piece is to get the correctly-formatted XML files working. The XML files are an advanced scenario with a UX that already isn't great so improvements are welcome, but really if we cared about the UX we would be designing a different solution that doesn't use these XML files in the first place. :)

Reply:
For now I'll leave it as is, and in the future I'll look into making xml schema errors a single error code in a separate PR.